tls/tls_ticket.c

```diff
@@ -608,6 +611,16 @@ ttls_ticket_parse(TlsCtx *ctx, unsigned char *buf, size_t len)
 	if (unlikely(r))
 		return r;
 
+	/*
+	 * If IP address of the client has changed, ask it to renew the
+	 * handshake. Otherwise a single open TLS session can be copied across
+	 * bots, and all the bots will renew the session. Allowed by OpenSSL,
+	 * acts as restoring session from file cache.
+	 */
+	if (tik->client_hash != ttls_cli_id_cb(ctx)) {
+		bzero_fast(tik, len);
+		return TTLS_ERR_SESSION_TICKET_EXPIRED;
+	}
```

Comment:
Please describe the behaviour in the TLS wiki as well as the motivation for the behaviour.

Also shouldn't we check whether current SNI matches to the ticket SNI and ask a client for the full handshake if they aren't match?

Reply:
First of all, currently we don't use the SNI for routing, and it possible to use different `host` headers in every single request in the same connection.  And at the moment, we don't support wildcard in SNI. 

> check the port from the Host and actual listening port.
>  We can do this if http_host_required is enabled 

Routing security is a much harder question that it seems. We have very rich `http_table` that controls forwarding. We also have two others sources for routing information: sticky sessions and SNI. And it's a hard question, which source must be used, because they all can give quite the opposite answers. I already raised this somewhere in the chat but we had no serious discussion on that. For me layered `http_hables` can be an answer here, like in netfilter hooks:
- pre-routing filtering: filters to block traffic, i.e. by origin, or by user-agent - it's expensive (due to header lookup for all incoming traffic) way to mitigate ddos. But this can be reasonable in some cases.
- routing. Here we can use all available sources to do vhost lookup: http_tables, SNI, cookies. We can optionally enable|disable some sources by configuration option. E.g. SNI is lower on OSI level so it can be not very helpful for, say, A-B testing.  There is no sense to prefer http_tables over cookies though.
- post-routing: variable check after vhost was defined.

We already have layered http_tables and the same approach can be implemented there. But other routing sources are not checked there and exist in parallel universe. Every time you're writing complicated set of tables and have sticky ookies enabled you have to worry what will happen with filters and routing. It's also an answer why the SNI is not used for routing now: it's on lower OSI level and can't know as much as tables do.

Reply:
> Also shouldn't we check whether current SNI matches to the ticket SNI and ask a client for the full handshake if they aren't match?

No we don't need to. I've designed ticket keys system to have this check fo free. Take a look how the ticket keys are generated, the key depends on secret defined by user, time and the host name. Secret is used as synchronising point, multiple Tempesta nodes using the shared secret can continue sessions opened on the other nodes. Time is used for key rotation. And the host name is used to give different keys for different vhosts. 

So if we can match requested vhost by SNI, we chose exact server certificate and ticket key from vhost description. Thus ticket borrowed by client from one SNI to another SNI can't be decrypted and used. The ticket is treated as expired in this case and fallback to full handshake is performed.

Reply:
Scenario with default `tls_certificate` directive was not covered, because in this case all tickets will be mapped to `default` vhost. Fixed by adding SNI hashsum into ticket value.